ppl/zqd/recruiter/README.md

```diff
@@ -0,0 +1,82 @@
+
+When zqd workers and root processes are started with the zqd listen command, they may be provided with the endpoint of the zqd recruit process.
+
+## API
+
+zqd recruit provides the following REST API:
+
+### /register
+
+{  
+"addr" : "*host:port for worker*",  
+"node" : "*ID of node in cluster*"  
+}
+
+/register is called by a worker process after it has started and is capable of processing /worker messages.
+
+/register will be called again when a zqd process completes a /worker request and is available to be recruited by another zqd root process for query execution. 
+
+When the recruiter process receives a /register request, it adds that worker to a pool of available workers. It maintains a list of available workers for each node in the cluster.
+
+In order to make the recruiter process state recoverable, /register will be called periodically when a  worker process is idle (e.g. not processing /worker requests) for a given timeout period. 
+This will allow a restarted zqd recruiter process to get registrations from running zqd workers.
+
+/register is a noop if the address is already registered, or if the worker is in the “reserved” pool. (See /unreserve below.) 
+
+### /recruit
+
+Request: {"N":*number of workers requested*}  
+Response: {"workers":[ *list of recruited workers addr,node* ]}
+
+/recruit is called by a zqd root process prior to starting query execution (i.e. /search). The number of workers returned may be less than requested, based on availability.
+
+/recruit implements a heuristic algorithm to minimize the number of recruited worker instances from a single node. For example, if it is possible to pick each worker from a separate node, /recruit will do so.
+
+When a worker is recruited, the worker is added to the “reserved pool” so it cannot be inadvertently re-registered and recruited again until it has been unreserved. See the /unreserve API below.
+
+### /unregister
+
+{"addr":"*host:port for worker*"}
+
+/unregister is called by a zqd worker process that gracefully terminates while “believing” that it is registered with the zqd recruiter.
+
+### /unreserve
+
+{"addr":"*host:port for worker*"}
+
+/unreserve is called by a zqd worker process that becomes idle after having completed a /worker request. It is removed from the reserved pool. This is a noop if it was not in the reserved pool.
+
+/unreserve is also called by a zqd worker process when it starts up, in case the addr of the worker was reserved by a previous process that is no longer running.
+
+The /unreserve API, along with the reserved pool maintained by /recruit, avoids the potential race condition of a worker re-registering itself after it has been recruited but before it has been sent work by a zqd root process.
+
+## Analysis of failures
```

Comment:
@marktwallace : The implementation described here differs from the 'long poll' idea we discussed (as a comment in the zqd recruit doc, and live). As written here, if a zqd root (or any other entity) halts after making a `/recruit` call, the workers would essentially need to be restarted to clear their reservations.

The goal of that 'long poll' idea was to connect the reservation state in the recruiter directly to the liveness of the zqd root: the root makes a connection to the recruiter, gets a reservation, and as long as it holds the connection open, the reservation is maintained. When it closes the connection, the reservation is released.  That simplifies the worker-recruiter needs as well: workers just ping the recruiter on startup, then periodically afterwards; the workers don't need to worry about their reservation state.

Reply:
A zqd worker will also clear the reserved state when it first becomes "idle" after finishing a /worker request, so a restart would not be needed. The client side of this system, including the connection implementation, is in progress in the branch recruiter-worker.